lib/client/ca_export.go

```diff
@@ -0,0 +1,289 @@
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		for _, ca := range cas {
+			if ca.GetClusterName() == localAuthName {
+				authorities = append(authorities, ca)
+			}
+		}
+	}
+
+	ret := strings.Builder{}
+	for _, ca := range authorities {
+		if req.ExportPrivateKeys {
+			for _, key := range ca.GetActiveKeys().SSH {
+				fingerprint, err := sshutils.PrivateKeyFingerprint(key.PrivateKey)
+				if err != nil {
+					return "", trace.Wrap(err)
+				}
+				if req.ExportAuthorityFingerprint != "" && fingerprint != req.ExportAuthorityFingerprint {
+					continue
+				}
+				ret.WriteString(string(key.PrivateKey))
+				ret.WriteString("\n")
+			}
+			continue
+		}
+
+		for _, key := range ca.GetTrustedSSHKeyPairs() {
+			fingerprint, err := sshutils.AuthorizedKeyFingerprint(key.PublicKey)
+			if err != nil {
+				return "", trace.Wrap(err)
+			}
+			if req.ExportAuthorityFingerprint != "" && fingerprint != req.ExportAuthorityFingerprint {
+				continue
+			}
+
+			// export certificates in the old 1.0 format where host and user
+			// certificate authorities were exported in the known_hosts format.
+			if req.CompatVersion == "1.0" {
+				castr, err := hostCAFormat(ca, key.PublicKey, client)
+				if err != nil {
+					return "", trace.Wrap(err)
+				}
+
+				ret.WriteString(castr)
+				ret.WriteString("\n")
+				continue
+			}
+
+			// export certificate authority in user or host ca format
+			var castr string
+			switch ca.GetType() {
+			case types.UserCA:
+				castr, err = userCAFormat(ca, key.PublicKey)
+			case types.HostCA:
+				castr, err = hostCAFormat(ca, key.PublicKey, client)
+			default:
+				return "", trace.BadParameter("unknown user type: %q", ca.GetType())
+			}
+			if err != nil {
+				return "", trace.Wrap(err)
+			}
+
+			// write the export friendly string
+			ret.WriteString(castr)
+			ret.WriteString("\n")
+		}
+	}
+
+	return ret.String(), nil
+}
+
+type exportTLSAuthorityRequest struct {
+	AuthType          types.CertAuthType
+	UnpackPEM         bool
+	ExportPrivateKeys bool
+}
+
+func exportTLSAuthority(ctx context.Context, client auth.ClientI, req exportTLSAuthorityRequest) (string, error) {
+	clusterName, err := client.GetDomainName(ctx)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	certAuthority, err := client.GetCertAuthority(
+		ctx,
+		types.CertAuthID{Type: req.AuthType, DomainName: clusterName},
+		req.ExportPrivateKeys,
+	)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	if len(certAuthority.GetActiveKeys().TLS) != 1 {
+		return "", trace.BadParameter("expected one TLS key pair, got %v", len(certAuthority.GetActiveKeys().TLS))
+	}
+	keyPair := certAuthority.GetActiveKeys().TLS[0]
+
+	ret := strings.Builder{}
+	marhsalKeyPair := func(data []byte) error {
+		if !req.UnpackPEM {
+			ret.WriteString(string(data))
+			ret.WriteString("\n")
+			return nil
+		}
+
+		b, _ := pem.Decode(data)
+		if b == nil {
+			return trace.BadParameter("no PEM data in CA data: %q", data)
```

Comment:
nit: Don't add data to the error message, it could be sensitive data.

(I realize we are exporting keys, but better safe than sorry.)